contracts/lockdrop/src/contract.rs

```diff
@@ -392,8 +455,435 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+        &Cw20QueryMsg::Balance {
+            address: env.contract.address.to_string(),
+        },
+    )?;
+    let current_balance = balance_response.balance;
+    attrs.push(attr("current_balance", current_balance.to_string()));
+
+    //update pool info
+    new_pool_info.amount_in_lockups = current_balance;
+    ASSET_POOLS_V2.save(deps.storage, pool_type, &new_pool_info)?;
+
+    // send stake message to generator
+    let stake_msgs = stake_messages(
+        config,
+        env.block.height + 1u64,
+        new_pool_info.lp_token,
+        current_balance,
+    )?;
+
+    MIGRATION_STATUS.save(deps.storage, &MigrationState::MigrateUsers(0u64))?;
+
+    Ok(Response::new()
+        .add_messages(stake_msgs)
+        .add_attributes(attrs))
+}
+
+fn migrate_users(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    limit: Option<u32>,
+) -> StdResult<Response> {
+    let mut attrs = vec![attr("action", "migrate_users")];
+    let migrate_state: MigrationState = MIGRATION_STATUS.load(deps.storage)?;
+    let limit = limit.unwrap_or(MIGRATION_USERS_DEFAULT_LIMIT);
+    let current_skip = MIGRATION_USERS_COUNTER
+        .may_load(deps.storage)?
+        .unwrap_or(0u32);
+
+    match migrate_state {
+        MigrationState::MigrateUsers(page) => {
+            let pool_types: Vec<PoolType> = ASSET_POOLS
+                .keys(deps.storage, None, None, Order::Ascending)
+                .collect::<Result<Vec<PoolType>, StdError>>()?;
+
+            let mut kfs: HashMap<String, Decimal256> = HashMap::new();
+            for pool_type in &pool_types {
+                let pool_info: PoolInfo = ASSET_POOLS.load(deps.storage, *pool_type)?;
+                let new_pool_info: PoolInfoV2 = ASSET_POOLS_V2.load(deps.storage, *pool_type)?;
+                kfs.insert(
+                    (*pool_type).into(),
+                    Decimal256::from_ratio(
+                        new_pool_info.amount_in_lockups,
+                        pool_info.amount_in_lockups,
+                    ),
+                );
+            }
+
+            let users: Vec<Addr> = USER_INFO
+                .keys(deps.storage, None, None, Order::Ascending)
+                .skip(current_skip as usize)
+                .take(limit as usize)
+                .collect::<Result<Vec<_>, _>>()?;
+            if users.is_empty() {
+                for pool_type in &pool_types {
+                    let mut pool_info: PoolInfo = ASSET_POOLS.load(deps.storage, *pool_type)?;
+                    let new_pool_info: PoolInfoV2 =
+                        ASSET_POOLS_V2.load(deps.storage, *pool_type)?;
+                    pool_info.amount_in_lockups = new_pool_info.amount_in_lockups;
+                    pool_info.lp_token = new_pool_info.lp_token;
+                    ASSET_POOLS.save(deps.storage, *pool_type, &pool_info, env.block.height)?;
+                }
+                MIGRATION_STATUS.save(deps.storage, &MigrationState::Completed)?;
```

Comment:
will we reach this code if we pass a 0 as limit?

Reply:
oh, that could be a disaster